ink-wrapper-types/src/calls.rs

```diff
@@ -0,0 +1,149 @@
+use std::marker::PhantomData;
+
+use ink_primitives::AccountId;
+
+/// Represents a call to a contract constructor.
+#[derive(Debug, Clone)]
+pub struct InstantiateCall<T: Send> {
+    /// The code hash of the contract to instantiate.
+    pub code_hash: [u8; 32],
+    /// The encoded data of the call.
+    pub data: Vec<u8>,
+    /// The salt to use for the contract.
+    pub salt: Vec<u8>,
+    /// The value to be sent with the call.
+    pub value: u128,
+    /// A marker for the type of contract to instantiate.
+    _contract: PhantomData<T>,
+}
+
+impl<T: Send> InstantiateCall<T> {
+    /// Create a new instantiate call.
+    pub fn new(code_hash: [u8; 32], data: Vec<u8>) -> Self {
+        Self {
+            code_hash,
+            data,
+            salt: vec![],
+            value: 0,
+            _contract: Default::default(),
+        }
+    }
+
+    /// Set the salt to use for the instantiation.
+    pub fn with_salt(mut self, salt: Vec<u8>) -> Self {
+        self.salt = salt;
+        self
+    }
+}
+
+pub struct InstantiateCallNeedsValue<T: Send> {
+    /// The code hash of the contract to instantiate.
+    pub code_hash: [u8; 32],
+    /// The encoded data of the call.
+    pub data: Vec<u8>,
+    /// A marker for the type of contract to instantiate.
+    _contract: PhantomData<T>,
+}
+
+impl<T: Send> InstantiateCallNeedsValue<T> {
+    pub fn new(code_hash: [u8; 32], data: Vec<u8>) -> Self {
+        Self {
+            code_hash,
+            data,
+            _contract: Default::default(),
+        }
+    }
+
+    pub fn with_value(self, value: u128) -> InstantiateCall<T> {
+        InstantiateCall {
+            value,
+            ..InstantiateCall::new(self.code_hash, self.data)
```

Comment:
oO
is this equivalent to
```rust
code_hash: self.code_hash,
data: self.data
}
```
?

Reply:
Only so long as `InstantiateCall` does not set any other defaults in the constructor. I chose to go this way instead of copying the fields over, because of that.